### DISCUSSION TITLE
International numbers

### DISCUSSION THREAD

**Discussion by cwrichardson** [vercel/next.js#42491] [Help]
Is there a native NextJS way to handle international number formatting? Or, barring a native way, what's the current best practice?

There was rather an issue-specific discussion around it that was initially an [issue](https://github.com/vercel/next.js/issues/37489) and got migrated to a [discussion](https://github.com/vercel/next.js/discussions/39425), but I haven't seen anything more general.

The issue is, if you're generating static pages, intl.numberFormat, if used directly, blows up all of the NextJS advantages. For example, taken from the NextJS docs and modified to use `intl`:

```
function Events({ events }) {
  return (
    <ul>
      {events.map((event) => (
        <li>{
            new Intl.DateTimeFormat('default', {
                year: '2-digit',
                month: 'numeric',
                day: 'numeric'
            }).format(new Date(event.date))
        }</li>
      ))}
    </ul>
  )
}

export async function getStaticProps() {
  const res = await fetch('https://.../events')
  const events = await res.json()

  return {
    props: {
      events,
    },
  }
}

export default Events
```

While that will look like it's working fine in development, and will build fine, when you ship it to production it blows up. Anyone who has a different system locale than whatever was on the server where you built it will get minified React errors [418](https://reactjs.org/docs/error-decoder.html/?invariant=418), [423](https://reactjs.org/docs/error-decoder.html/?invariant=423), and [425](https://reactjs.org/docs/error-decoder.html/?invariant=425).* The page will still display "fine", but it will have fully re-rendered on the client from the root, so you've lost all the benefit of having generated a static page in the first place.

You can work around this, by putting all of your numbers inside of useEffect, as mentioned in the previous discussion. That might look something like this:

```
function Events({ events }) {

  useEffect(() => {
      events.forEach((event, i) => {
          this[i].formattedDate = new Intl.DateTimeFormat('default', {
                year: '2-digit',
                month: 'numeric',
                day: 'numeric'
            }).format(new Date(event.date));
      }, arr);
  }, []);

  return (
    <ul>
      {events.map((event) => (
        <li>{event.formattedDate}</li>
      ))}
    </ul>
  )
}
```

That would fix the client-side errors, and prevent it from rendering from the root on the client, but this is already not ideal. Then, take into account that maybe it's not that simple of a use case. Maybe instead of dates, we have prices, and need to conditionally apply a discount. But now your prices won't update, because they're in a useEffect hook, and all of a sudden you're turning every number into a ref:

```
const EventPurchaseCard = (props) => {
    const {
        discountAmount = 0,
        discountEndDate = null,
        event
    } = props.details;

    const discountEndShort = useRef();
    const eventDate = useRef();
    const discount = useRef();
    const formattedPrice = useRef();
    const formattedOriginalPrice = useRef();
    const price = useRef();

    useEffect(() => {
        price.current = event.price;

        if ((discountEndDate && (new Date(discountEndDate) > new Date())) {
            price.current = Math.max(0, price.current - discountAmount);
        }

        if (discountEndDate) {        
            discountEndShort.current = new Intl.DateTimeFormat('default', {
                year: '2-digit',
                month: 'numeric',
                day: 'numeric'
            }).format(new Date(discountEndDate));
        }

        if (event?.dateTime) {
            eventDate.current = new Intl.DateTimeFormat('default', {
                weekday: 'long',
                year: 'numeric',
                month: 'long',
                day: 'numeric',
                hour: 'numeric',
                minute: 'numeric'
            }).format(new Date(event.dateTime));
        }

        formattedPrice.current = new Intl.NumberFormat('default', {
            style: 'currency',
            currency: currency,
            currencyDisplay: 'symbol',
            minimumFractionDigits: 0,
            maximumFractionDigits: 2
        }).format(price.current);

        ...
    }, []);
    ...
```

This **has to be** some kind of anti-pattern.

I'm hoping NextJS has a native way of handling this, but all I can find in the docs is about internationalisation of routes.

I have it working in my current project, but in addition to being gross, it's not performing well. My mobile PageSpeed metrics were in the mid-90s for mobile, which is rather part of the magic of NextJS (at least in my opinion). Now they're in the mid 70s.

How are other people solving this?

NB: I also posted this on [stack overflow](https://stackoverflow.com/questions/74316543/international-number-formats-in-nextjs), but figured I should come to the source, so to speak :)

**Comment by icyJoseph** [comment]
So this is an issue with hydration errors being problematic within concurrent rendering in React 18.

An escape-hatch is to use https://reactjs.org/docs/dom-elements.html#suppresshydrationwarning (supressHydrationWarning), directly on the DOM element that's bound to have some kind of divergence. Do not overuse this, exhaust every other possibility to eliminate the hydration issue first.

A more elaborate trick is to render a sensible default on the server, and send that to the client. You could even make this invisible with CSS, so that it keeps its space while being invisible (display none and then block, would cause layoutshift). Once you are in the client you would update the text node with a useEffect and make the element visible with CSS>
